### ISSUE TITLE
Slimming down SharpDX (The main assembly, not the project)

### ISSUE THREAD

**Issue by PathogenDavid** [sharpdx/SharpDX#398]
First, I apologize if this has been discussed before or there is already an ongoing effort to do what I am describing. I could not find anything related to it.

Second, I am mostly posting this because I am making these changes to a personal version of SharpDX anyway. I thought I should start a discussion to determine if I should make my changes in a manner that allows them to be integrated into the main SharpDX repository at some point. I realize that what I am proposing is somewhat of a breaking change, but I thought I should say something in case Alexandre is already working on something similar at Silicon Studio or there is a general interest in this being done.

---

SharpDX advertises itself as "an open-source project delivering the full DirectX API under the .Net platform". While SharpDX certainly accomplishes this goal, it also comes with a ton of utility code to facilitate easier development of graphical applications. Most of this is the SharpDX Toolkit, but there is still much of it in the main SharpDX assemblies.

This unnecessary code seems to come from three primary locations:
1. Extra math stuff that was brought in when the SlimDX math library was added. These are types that may be useful for 3D math, but are never used by any of the binding assemblies. (Example: Angle)
2. Utility code that serves narrow internal uses for Toolkit. (Example: TaskUtil)
3. General utility code to make certain things easier (Example: RandomUtil)

As an example, the following files are unnecessary for building and using the SharpDX bindings:

```
Angle.cs
BoundingBox.cs
BoundingFrustum.cs
BoundingSphere.cs
Collections/ObservableCollection.cs
Collections/ObservableCollectionEventArgs.cs
Collections/ObservableDictionary.cs
Collections/ObservableDictionaryEventArgs.cs
Collision.cs
ComBaseStreamNative.cs
ComponentCollection.cs
DataBuffer.cs
Diagnostics/CollectionDebugView.cs
Direct3D/PixHelper.cs
Enums.cs
FrustumCameraParams.cs
Half.cs
Half2.cs
Half3.cs
Half4.cs
HalfUtils.cs
IO/NamespaceDoc.cs
IO/PathUtility.cs
IServiceRegistry.cs
IdentityEqualityComparer.cs
MathUtil.cs
Matrix3x3.cs
Multimedia/RiffChunk.cs
Multimedia/RiffParser.cs
Multimedia/SoundStream.cs
Multimedia/SpeakersExtensions.cs
Multimedia/WavWriter.cs
OrientedBoundingBox.cs
Plane.cs
RandomUtil.cs
Ray.cs
Serialization/ArrayLengthType.cs
Serialization/BinarySerializer.cs
Serialization/DynamicSerializerAttribute.cs
Serialization/IDataSerializable.cs
Serialization/InvalidChunkException.cs
Serialization/NamespaceDoc.cs
Serialization/SerializeFlags.cs
Serialization/SerializerMode.cs
ServiceEventArgs.cs
SingletonString.cs
Threading/TaskUtil.cs
TimerTick.cs
Windows/NamespaceDoc.cs
Windows/RenderControl.cs
Windows/RenderForm.cs
Windows/RenderLoop.cs
```

There may be more, but these are what I found while slimming down SharpDX for my own project. I more or less determined this through a combination of Visual Studio's "Find all references" tool, and removing+re-adding things while building for every target platform. I also removed much of the inter-dependencies between math library bits and anything using the serialization simply for the purpose of making things bare bones.

---

My proposal would be to:
- Move all math-related structs/classes into a SharpDX.Math assembly. (Potentially continuing to use the SharpDX namesapce to prevent breaking applications father than a missing library reference.)
- Move serialization into separate assembly. (This might not be super realistic, but it'd be nice to include a DirectX binding library without also getting a serialization library.)
- Move Direct3D utilities into separate assembly. (Keep graphics stuff separate.)
- Move Multimedia classes into separate assembly. (Keep audio stuff separate.)
- Move all general utility stuff into SharpDX.Toolkit.Utilities
- Move things more appropriate for toolkit (SharpDX.Windows) to new Toolkit library.
- Separate Toolkit into entirely separate repository (Similar to how the samples are completely separate.)

My personal primary motivations for this are:
- To further the consistency of the SharpDX/Toolkit division
- Remove unecessary baggage from the core assembly. (So, for example, you can use DirectX11 without an assembly that also includes a bunch of audio stuff.)
- Enhance the separation of interests in the codebase.
- Make it so a project can integrate SharpDX more easily without using the SharpDX math library. (This is my biggest motivation, the primary benefit of this is for cross-platform game engines that don't want to rely on another math implementation. Unfortunately, C# doesn't really let you fudge this by using hacky casts like SharpDX::Color\* c = (SharpDX::Color_)(void_)&myColorThatHasSameMemoryLayout; so no "free" casts.)

I guess the overall goal is to make SharpDX more of a pure DirectX managed binding, rather than the "DirectX managed binding (And more!)" that it is right now.

---

Once again, I am not posting this issue because I am trying to tell you how to maintain your library. I am posting it because if it is something the primary maintainers desire, I am willing to put more effort into "doing it right" rather than making my own custom fork of SharpDX.

TL;DR: The main SharpDX assembly nets you a lot of unnecessary cruft, I am reducing it for my own purposes and want to know if there is interest in this being something done in the main repo.
